pallets/asset-index/src/lib.rs

```diff
@@ -677,6 +717,43 @@ pub mod pallet {
 			share.checked_div(&nav).ok_or_else(|| ArithmeticError::Overflow.into())
 		}
 
+		/// Consolidate and clean deposits while completing withdraw
+		fn do_consolidate_deposits(caller: &T::AccountId, mut amount: T::Balance) -> Result<T::Balance, DispatchError> {
+			<Deposits<T>>::try_mutate_exists(&caller, |maybe_depositing| -> Result<T::Balance, DispatchError> {
+				let mut depositing = maybe_depositing.take().ok_or(<Error<T>>::NoDeposits)?;
+				let mut total_fee: T::Balance = Default::default();
+
+				let mut dim: Option<(T::Balance, T::BlockNumber)> = None;
+				depositing.retain(|(index_tokens, block_number)| {
+					if amount >= *index_tokens {
+						amount -= *index_tokens;
+						total_fee += T::RedemptionFee::redemption_fee(*block_number, *index_tokens);
+						false
+					} else {
+						if *index_tokens > amount {
+							dim = Some((*index_tokens - amount, *block_number));
+							total_fee += T::RedemptionFee::redemption_fee(*block_number, amount);
+							amount = Default::default();
```

Comment:
We can clean this up a bit,
```
*index_tokens = *index_tokens - amount;
```
Then there is also no need for the `dim` at all, right?

Reply:
`retain` provides a reference which we can not modify it

Reply:
I see, makes sense, then we only need to store `let mut rem: Option<T::Balance>` and update this like you did already